victor/src/layout/boxes/generation.rs

```diff
@@ -31,174 +27,433 @@ struct Context<'a> {
+                BlockContainer::BlockLevelBoxes(
+                    intermediate_block_levels
+                        .into_par_iter()
+                        .map(|block_level| block_level.finish(context))
+                        .collect(),
+                )
+            }
+        })
+    }
 }
 
-impl PushBlock for BlockContainerBuilderExtra {
-    fn push_block(builder: &mut Builder<Self>, block: BlockLevel) {
-        if !builder.consecutive_inline_levels.is_empty() {
-            builder.wrap_inlines_in_anonymous_block();
+impl IntermediateBlockLevelBox {
+    fn finish(self, context: &Context) -> BlockLevelBox {
+        match self {
+            IntermediateBlockLevelBox::SameFormattingContextBlock { style, contents } => {
+                BlockLevelBox::SameFormattingContextBlock {
+                    contents: contents.finish(context, &style),
+                    style,
+                }
+            }
         }
-        builder.extra.block_levels.push(block)
     }
 }
-impl Builder<BlockContainerBuilderExtra> {
-    fn wrap_inlines_in_anonymous_block(&mut self) {
-        self.extra
-            .block_levels
-            .push(BlockLevel::SameFormattingContextBlock {
-                style: ComputedValues::anonymous_inheriting_from(&self.style),
-                contents: BlockContainer::InlineFormattingContext(
-                    self.consecutive_inline_levels.take(),
-                ),
-            });
-    }
 
-    fn build(mut self) -> (Arc<ComputedValues>, BlockContainer) {
-        if !self.consecutive_inline_levels.is_empty() {
-            if self.extra.block_levels.is_empty() {
-                return (
-                    self.style,
-                    BlockContainer::InlineFormattingContext(self.consecutive_inline_levels),
+impl IntermediateBlockContainer {
+    fn finish(self, context: &Context, style: &Arc<ComputedValues>) -> BlockContainer {
+        match self {
+            IntermediateBlockContainer::BlockBox(block) => {
+                IntermediateBlockFormattingContextBuilder::new_for_descendant(
+                    context,
+                    block,
+                    style.clone(),
                 )
+                .build()
+                .finish(context)
+                .0
             }
-            self.wrap_inlines_in_anonymous_block()
+            IntermediateBlockContainer::InlineFormattingContext(
+                intermediate_inline_formatting_context,
+            ) => BlockContainer::InlineFormattingContext(
+                intermediate_inline_formatting_context.finish(context),
+            ),
+        }
+    }
+}
+
+impl IntermediateInlineFormattingContext {
+    fn finish(self, context: &Context) -> InlineFormattingContext {
+        InlineFormattingContext {
+            inline_level_boxes: self.inline_level_boxes,
+            text_runs: self
+                .text_runs
+                .into_par_iter()
```

Comment:
This is fine for now, but we’ll probably have enough parallelism between separate paragraphs (blocks) that we don’t need to parallelize shaping within a paragraph (one inline formatting context). In fact doing complex shaping correctly (including across inline element boundaries) might require us to sequentialize. Though separating what this module calls `build` and `finish` phases might still be useful so that shaping can have the context of the entire paragraph, so this isn’t wasted effort ;)

(But shaping is for [Skribo](https://raphlinus.github.io/rust/skribo/text/2019/02/27/text-layout-kickoff.html) to solve, naive shaping will do for CSS layout prototyping.)